File: app/journal_asset_import.ml
module Ui = Bonsai_swiftui_ui
module Uuid = Logseq_db_types.Graph_types.Uuid

let decode ~target payload =
  let ( let* ) = Result.bind in
  try
    let json = Yojson.Basic.from_string payload in
    let field key =
      match Yojson.Basic.Util.member key json with
      | `String value -> Ok value
      | _ -> Error "Invalid attachment selection"
    in
    let uuid key =
      let* text = field key in
      Result.map_error (fun _ -> "Invalid import identity") (Uuid.of_string text)
    in
    let* operation = uuid "operation" in
    let* asset = uuid "asset" in
    let* local_mutation = uuid "localMutation" in
    let* metadata_mutation = uuid "metadataMutation" in
    let* replace_reference =
      match json with
      | `Assoc fields ->
        (match List.assoc_opt "replaceReference" fields with
         | Some `Null -> Ok None
         | Some (`String id) -> Result.map Option.some (Uuid.of_string id)
         | _ -> Error "Invalid replacement reference")
      | _ -> Error "Invalid attachment selection"
    in
    let* source_file = field "path" in
    let* title = field "title" in
    let* file_type = field "type" in
    Ok
      Logseq_db_types.Asset_import.
        { operation
        ; asset
        ; target
        ; replace_reference
        ; local_mutation
        ; metadata_mutation
        ; source_file
        ; title
        ; file_type
        }
  with
  | _ -> Error "Invalid attachment selection"
;;

let extension =
  Ui.Native_widget.Extension.create
    ~kind_id:(Bonsai_swiftui_spec.Id.Native_widget.Kind_id.of_int 2104)
    ~version:1
    ~capabilities:[ Stateful; Resource; Semantics ]
    ~encode_props:(fun props -> Bytes.of_string (Yojson.Basic.to_string props))
    ~decode_event:(fun ~event_id:_ bytes -> Ok (Bytes.to_string bytes))
    ()
;;

let is_dismissal payload =
  match (try Yojson.Basic.from_string payload with _ -> `Null) with
  | `Assoc fields ->
    (match List.assoc_opt "action" fields with
     | Some (`String "dismissed") -> true
     | _ -> false)
  | _ -> false
;;

let view ~key ~enabled ~completion ~replacement ~request ~on_select =
  let operation, error =
    match completion with
    | None -> `Null, `Null
    | Some (operation, error) ->
      ( `String operation
      , (match error with
         | None -> `Null
         | Some message -> `String message) )
  in
  Ui.Native_widget.widget
    extension
    ~key
    ~props:
      (`Assoc
          [ "enabled", `Bool enabled
          ; "completion", operation
          ; "error", error
          ; ( "replace"
            , match replacement with
              | None -> `Null
              | Some root -> `String root )
          ; "request", `Int request
          ])
    ~on_event:on_select
    ~children:[]
    ()
;;

File: swift/JournalAssetImport.swift
import LUIAppleBackend
import Observation
import SwiftUI
import UniformTypeIdentifiers

@MainActor enum JournalAssetImport {
  struct Properties: Decodable {
    let enabled: Bool
    let completion: String?
    let error: String?
    let replace: String?
    let request: Int
  }

  @Observable final class Selection {
    var operation: String?
    private var source: URL?
    private var scoped = false

    func retain(_ url: URL, operation: String) {
      release()
      source = url
      scoped = url.startAccessingSecurityScopedResource()
      self.operation = operation
    }

    func release() {
      if scoped { source?.stopAccessingSecurityScopedResource() }
      source = nil
      scoped = false
      operation = nil
    }
  }

  struct View: SwiftUI.View {
    let context: LUIAppleExtensionViewContext
    @State private var selection = Selection()
    @State private var presented = false
    @State private var handled = false
    @State private var error: String?

    private var properties: Properties? {
      JournalExtensions.decode(Properties.self, context: context)
    }

    private func emitDismissed() {
      guard !handled,
        let data = try? JSONSerialization.data(withJSONObject: ["action": "dismissed"])
      else { return }
      handled = true
      JournalExtensions.emit(context: context, payload: data)
    }

    var body: some SwiftUI.View {
      Button {
        guard context.isUserInteractionEnabled else { return }
        presented = true
      } label: {
        Label(selection.operation == nil ? "Attach file" : "Importing file", systemImage: "paperclip")
          .labelStyle(.iconOnly)
      }
      .disabled(properties?.enabled == false || selection.operation != nil)
      .accessibilityIdentifier("journal-asset-import")
      .fileImporter(isPresented: $presented, allowedContentTypes: [.item], allowsMultipleSelection: false) { result in
        guard context.isUserInteractionEnabled else { return }
        do {
          guard let source = try result.get().first else {
            if properties?.replace != nil { emitDismissed() }
            return
          }
          handled = true
          let operation = UUID().uuidString.lowercased()
          selection.retain(source, operation: operation)
          let extensionName = source.pathExtension.lowercased()
          let payload = try JSONSerialization.data(withJSONObject: [
            "operation": operation, "asset": UUID().uuidString.lowercased(),
            "localMutation": UUID().uuidString.lowercased(), "metadataMutation": UUID().uuidString.lowercased(),
            "path": source.path(percentEncoded: false), "title": source.lastPathComponent,
            "replaceReference": properties?.replace ?? NSNull(),
            "type": extensionName.isEmpty ? "bin" : extensionName,
          ] as [String: Any])
          if !JournalExtensions.emit(context: context, payload: payload) {
            selection.release()
            error = "The destination is no longer available. Select the file again."
          }
        } catch {
          selection.release()
          if (error as NSError).code == NSUserCancelledError {
            if properties?.replace != nil { emitDismissed() }
          } else {
            self.error = "Unable to access the selected file. Please try again."
          }
        }
      }
      .onChange(of: presented) { _, isPresented in
        if isPresented {
          handled = false
        } else if properties?.replace != nil {
          // iOS never invokes the fileImporter completion on Cancel, so treat
          // closing an armed picker without a pick as a dismissal.
          emitDismissed()
        }
      }
      .onChange(of: properties?.request) { _, _ in
        if properties?.replace != nil { presented = true }
      }
      .onChange(of: properties?.completion) { _, operation in
        guard let operation, operation == selection.operation else { return }
        selection.release()
        error = properties?.error
      }
      .onDisappear { selection.release() }
      .alert("Unable to import file", isPresented: Binding(get: { error != nil }, set: { if !$0 { error = nil } })) {
        Button("OK", role: .cancel) { error = nil }
      } message: { Text(error ?? "") }
    }
  }
}
